### ISSUE TITLE
Method `route()` does not always trigger `onChange` prop of `Router` component

### ISSUE THREAD

**Issue by sydev** [developit/preact-router#291]
When I redirect to a specific route like `route('/')` it doesn´t trigger the `onChange` prop of the `Router` component. Here´s my code:

```javascript
import { h, Component } from 'preact';
import { Router } from 'preact-router';

import * as um from '../lib/user-management';

import Header from './Header';
import Footer from './Footer';

import Login from 'async!../routes/login';
import Logout from 'async!../routes/logout';
import ImageTool from 'async!../routes/image-tool';
import Category from 'async!../routes/category';

export default class App extends Component {

  state = {
    showHeaderMenu: false
  }

  /** Gets fired when the route changes.
   *  @param {Object} event    "change" event from [preact-router](http://git.io/preact-router)
   *  @param {string} event.url  The newly routed URL
   */
  handleRoute = async e => {
   try {
      const isLoggedIn = await um.isLoggedIn();
      if (!isLoggedIn) return route('/login');

      const token = await um.getToken();
      const valid = await um.verifyToken(token);
      if (!valid) return route('/login');

      this.currentUrl = e.url;

      if (e.url.indexOf('/login') >= 0) this.setState({showHeaderMenu: false});
      else this.setState({showHeaderMenu: true});
    } catch (err) {
      console.error(err);
    }
  };

  render(props, state) {
    return (
      <div id="app">
        <Header showMenu={ state.showHeaderMenu } />
        <Router onChange={ this.handleRoute }>
          <ImageTool path="/" />
          <Login path="/login" />
          <Logout path="/logout" />
          <Category path="/category/:id" />
        </Router>
        <Footer />
      </div>
    );
  }
}
```

Am I doing something wrong?

**Comment by ConradSollitt** [comment]
Just came across a very similar issue so it might be related and if they are I know what the issue is.

I'm developing online without webpack and using the following versions of Preact, Preact Router, and related:

```html
<script src="https://unpkg.com/preact@10.5.7/dist/preact.min.js"></script>
<script src="https://unpkg.com/preact-router@3.2.1/dist/preact-router.js"></script>
<script src="https://unpkg.com/preact-router@3.2.1/match.js"></script>
<script src="https://unpkg.com/history@5.0.0/umd/history.production.min.js"></script>
```
Here is the JSX I was testing:

```jsx
<Router history={createHashHistory()} onChange={(e) => console.log(e)}>
    <HomePage path="/" />
    <DataPage path="/data" />
</Router>
```

When I debugged using Chrome DevTools I found that `onChange` is being replaced by `onInput` for Preact Router (screenshots below).

**Reason for the issue**
https://preactjs.com/guide/v10/differences-to-react/#use-oninput-instead-of-onchange

My initial thought is `onChanged` could be replaced by a new API `onRouteChanged` to avoid the issue. The event name is a little longer but the name clearly defines what the event is handling.

Additionally I was debugging this and found this topic because I was finding `<Link>` from `preact-router/match` does not set `activeClassName` when the page loads, only after the user changes the route (at least in the app I'm working on). My solution was to use the `onChange` to handle it the first time manually.

![image](https://user-images.githubusercontent.com/57777521/101978497-3e731100-3c0a-11eb-923e-ef11c45314d7.png)

![image](https://user-images.githubusercontent.com/57777521/101978531-7712ea80-3c0a-11eb-9225-7a531fbb44bf.png)

**Comment by ConradSollitt** [comment]
For reference, here is the final version of the page I was testing (full source including JSX is on the page as JSX is compiled in the browser).

https://www.dataformsjs.com/getting-started/en/template-preact-router.htm

Rather than handling `onChange` from React Router for the active link I ended up using `<Link>` from `preact-router@/match.js` I use the regular `<Link>` from `preact-router.js` and manually set the links outside of rendering. Arguably not an ideal solution but it works correctly in all browsers.

```jsx
// After the main render
preact.render(<App />, document.getElementById('root'));

// I Use `preactRouter.subscribers` and `window.setTimeout`
function setActiveNavLink() { /* ... */ }
window.preactRouter.subscribers.push(setActiveNavLink); // Route Change
window.setTimeout(setActiveNavLink, 100); // Page Initial Load
```

**Comment by Shekesh** [comment]
I'm having the same issue. I have a router with only a child component with `default`. Using `route()` in the handler for `onChange` has no effect. The only solution I found so far is described in [this issue](https://github.com/preactjs/preact-router/issues/302): Using `route` inside `setTimeout` which in my opinion isn't the most elegant way. I also would expect that `route()` inside the `onChange`-handler would work.
Any updates on this?

**Comment by motss** [comment]
Pretty new to Preact and encountered such issue but these are my findings between `route()` and `e.router.routeTo()`.

> TLDR; ROUTES is empty when `route()` is called.

```ts
1   const handleChange: RouterProps['onChange'] = (e) => {
2     switch (e.url) {
3       case '/user': {
4         route('/401');  // setTimeout(() => route('/401'), 3e3);
5         e.router.routeTo('/401');
6         break;
7      }
8      default:
9     }
10  };
```

1. `route()` at L4 - it does not work because by the time `handleChange` gets triggered [ROUTERS](https://github.com/preactjs/preact-router/blob/0401c31559be55d8a15aaf513aa7a941a766f42b/src/index.js#L12) has yet to be populated with all predefined route objects which renders `route()` to always return false. This can be resolved by using `setTimeout()` but it feels hacky.
3. `e.router.routeTo()` at L5 - this works because it depends on the children within `<Router />` instead of `ROUTES` to do any checks such as `canRoute(...)`.

**Comment by pnutmath** [comment]
Still facing same issue, any update?
